### PR TITLE
feature: support HTTP/3 testing.

### PR DESCRIPTION
**Pull request by anguslee**

### CODE REVIEW

File: lib/Test/Nginx/Util.pm [hunk 1]
```diff
@@ -2688,6 +2728,12 @@ sub use_http2 ($) {
             return undef;
         }
 
+        if (defined $block->http3) {
+            warn "WARNING: ", $block->name, " - explicitly rquires HTTP/3, so will not use HTTP/2\n";
+            $block->set_value("test_nginx_enabled_http0", 0);
```
Comment (xiaocang) [inline lib/Test/Nginx/Util.pm:2733]: `http0` is typo here?

File: lib/Test/Nginx/Util.pm [hunk 2]
```diff
@@ -78,6 +81,19 @@ our $ServerAddr = '127.0.0.1';
...
 our $ServerName = 'localhost';
 
+our $Http3SSLCrt = $ENV{TEST_NGINX_HTTP3_CRT};
+our $Http3SSLCrtKey = $ENV{TEST_NGINX_HTTP3_CRT_KEY};
+
+if ($UseHttp2 && $UseHttp3) {
+    die "Ambiguous: both TEST_NGINX_USE_HTTP3 and TEST_NGINX_USE_HTTP2 are set.\n";
+}
+
+if ($UseHttp3 && (!defined $Http3SSLCrt || !defined $Http3SSLCrtKey)) {
```
Comment (xiaocang) [inline lib/Test/Nginx/Util.pm:91]: We shouldn't force the user to set the path to the certificate, they can load the certificate in other ways: perhaps by modifying the `http_config` or `config` block, perhaps by loading the certificate by lua

File: lib/Test/Nginx/Util.pm [hunk 3]
```diff
@@ -78,6 +81,19 @@ our $ServerAddr = '127.0.0.1';
...
+    die "Must set both env TEST_NGINX_HTTP3_CRT and TEST_NGINX_HTTP3_CRT_KEY in global HTTP/3 test mode\n";
+}
+
+our $ServerConfigHttp3 = '';
```
Comment (xiaocang) [inline lib/Test/Nginx/Util.pm:95]: Maybe we don't need this variable: setting up tls version and certificate path should be done by the user himself?
cc @agentzh

**Review by xiaocang — CHANGES_REQUESTED** [review]

File: lib/Test/Nginx/Util.pm (hunk 2 above)
Comment (anguslee) [reply lib/Test/Nginx/Util.pm:91]: This feature allows the user to run all tests in HTTP/3 without modifying individual .t test case files.

File: lib/Test/Nginx/Socket.pm [hunk 4]
```diff
@@ -1807,7 +1807,21 @@ sub send_request ($$$$@) {
...
+        #warn "Request: $req\n";
+        if ($req =~ /^\s*HEAD\s+/) {
+            #warn "Found HEAD request!\n";
+            $head_req = 1;
+        }
+    }
+
+    if (use_http2($block) || use_http3($block)) {
+        return send_http_req_by_curl($block, $req, $timeout), $head_req;
+    }
```
[the remark this answers is not in the document]
Comment (anguslee) [reply lib/Test/Nginx/Socket.pm:1824]: HTTP/3 servers listen on UDP sockets only.

[the remark this answers is not in the document]
Comment (xiaocang) [reply lib/Test/Nginx/Socket.pm:1824]: here's the document from nginx

> how does the client know which type of connection to request initially, TCP or UDP? The solution is for the client to establish a TCP connection for the initial HTTP request

https://www.nginx.com/blog/introducing-technology-preview-nginx-support-for-quic-http-3/

**Pull request closed without merging**